### PR TITLE
Extra spline utility functions

### PR DESCRIPTION
**Pull request by freddyaboulton**
Adds the following methods to the arcline_spline_utils module:

- length_of_lane
- project_pose_to_lane
- get_curvature_at_distance_along_lane

Adds the following methods to the map api:
- get_lane . Gets the arcline path representation of the lane given its token. 
- find_closest_lane. Find the ID of the closest lane given an x y point.

PR #304 Should be merged first

### CODE REVIEW

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py [hunk 1]
```diff
@@ -21,7 +25,7 @@ def principal_value(angle_in_radians: float) -> float:
     return scaled_angle
 
 
-def compute_segment_sign(arcline_path_3: Dict[str, Any]) -> Tuple[int, int, int]:
+def compute_segment_sign(arcline_path_3: ArcLinePath) -> Tuple[int, int, int]:
```
Comment (emwolff) [inline python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Let's be consistent with the name. Maybe just `ArcLinePath`? We can remove the 3 to clarify the name, IMO.

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py [hunk 2]
```diff
@@ -103,7 +107,7 @@ def _get_lie_algebra(segment_sign: Tuple[int, int, int],
 
 
 
-def pose_at_length(arcline_path_3: Dict[str, Any],
+def pose_at_length(arcline_path_3: ArcLinePath,
                    l: float) -> Tuple[float, float, float]:
     """
     Retrieves pose at step meters along the arcline_path_3.
```
Comment (emwolff) [inline python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: `step` does not exist.

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py [hunk 3]
```diff
@@ -186,19 +190,97 @@ def discretize(arcline_path_3: Dict[str, Any],
...
+    for path in lane:
+        poses = discretize(path, resolution_meters)
         for pose in poses:
             pose_list.append(pose)
     return pose_list
+
+
+def length_of_lane(lane: List[ArcLinePath]) -> float:
+    """
+    Calculates the length of a lane
```
Comment (emwolff) [inline python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Need to add units to doc.

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py [hunk 4]
```diff
@@ -186,19 +190,97 @@ def discretize(arcline_path_3: Dict[str, Any],
     return discretization
 
 
-def discretize_lane(arcline_list: List[Dict[str, Any]],
+def discretize_lane(lane: List[ArcLinePath],
                     resolution_meters: float) -> List[Tuple[float, float, float]]:
```
Comment (emwolff) [inline python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Can we use Pose type for the return hint?

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py [hunk 5]
```diff
@@ -4,9 +4,13 @@
 import math
 from typing import Dict, Any, List, Tuple
 
+import numpy as np
+
 # (x, y, yaw) in global frame
 Pose = Tuple[float, float, float]
 
+ArcLinePath = Dict[str, Any]
```
Comment (emwolff) [inline python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Nice! Will clarify the interface and we can strengthen the type later.

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py [hunk 6]
```diff
@@ -186,19 +190,97 @@ def discretize(arcline_path_3: Dict[str, Any],
...
+    """
+    Find the closest pose on a lane to a query pose and additionally return the
+    distance along the lane for this pose. Note that this function does
+    not take the heading of the query pose into account.
+    :param pose: Query pose
+    :param lane: Will find the closest pose on this lane
+    :return: Tuple of the closest pose and the distance along the lane
+    """
+
+    discretized_lane = discretize_lane(lane, resolution_meters=0.5)
```
Comment (emwolff) [inline python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Let's let the user control the resolution. Say make it a default param?

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py [hunk 7]
```diff
@@ -186,19 +190,97 @@ def discretize(arcline_path_3: Dict[str, Any],
...
+
+    closest_pose = discretized_lane[closest_pose_index]
+    distance_along_lane = closest_pose_index * 0.5
+    return closest_pose, distance_along_lane
+
+
+def _find_index(distance_along_lane: float, lengths: List[float]) -> int:
+    """
+    Helper function for finding of path along lane corresponding to the distance_along_lane
+    :param distance_along_lane: Distance along the lane
```
Comment (emwolff) [inline python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Units in doc.

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py [hunk 8]
```diff
@@ -186,19 +190,97 @@ def discretize(arcline_path_3: Dict[str, Any],
...
+
+    if len(lengths) == 1:
+        return 0
+    else:
+        return min(index for index, length in enumerate(lengths) if distance_along_lane <= length)
+
+
+def get_curvature_at_distance_along_lane(distance_along_lane: float, lane: List[ArcLinePath]) -> float:
+    """
+    Computes the unsigned curvature at a distance along a lane
```
Comment (emwolff) [inline python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Units in doc.

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py [hunk 9]
```diff
@@ -186,19 +190,97 @@ def discretize(arcline_path_3: Dict[str, Any],
...
+    :param distance_along_lane: Distance along the lane to calculate the curvature at
+    :param lane: Lane to query
+    :return: Curvature, always non negative.
+    """
+
+    segment_lengths = np.cumsum([sum(path['segment_length']) for path in lane])
```
Comment (emwolff) [inline python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Is the name consistent with what you are computing? It is the length from start to end of each segment.

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py (hunk 1 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Good idea!

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py (hunk 2 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Good catch!

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py (hunk 3 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Done!

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py (hunk 4 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Yep, I missed that.

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py (hunk 7 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Done

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py (hunk 8 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Done!

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py (hunk 9 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: You're right! Changed the name to total_length_at_segment

File: python-sdk/nuscenes/map_expansion/arcline_path_utils.py (hunk 6 above)
Comment (freddyaboulton) [reply python-sdk/nuscenes/map_expansion/arcline_path_utils.py]: Sure thing.

**Comment by freddyaboulton** [comment]
@emwolff Addressed your comments!

**Comment by freddyaboulton** [comment]
Closing in favor of #309

**Pull request closed without merging**